Initial go client

dalehurwitz opened the pull request.

Commit:
init

Commit:
Basic Go client with sms & email send

jonathaningram commented on README.md:
````diff
@@ -1 +1,223 @@
+## Getting started
+
+```Go
+client, err := notify.NewClient(api_key)
````
@dalehurwitz `apiKey` (can also change any Go variables named as such)

jonathaningram commented on README.md:
````diff
@@ -1 +1,223 @@
+
+```go
+resp, err := client.SendSMS(
+  templateId,
````
`templateID` (idiomatic Go)

jonathaningram commented on README.md:
```diff
@@ -1 +1,223 @@
+resp, err := client.SendSMS(
+  templateId,
+  phoneNumber,
+  notify.Reference("Sam's reminders"),
```
Reference would usually be an internal reference number, so maybe better as more of an ID e.g. `1234`.

jonathaningram replied:
I see that it can also be used for batches. Probably OK to leave it as you have.

jonathaningram reviewed (commented):
Sending over pending comments. Haven't finished reviewing!

Commit:
var names

jonathaningram commented on options.go:
```diff
@@ -0,0 +1,104 @@
+	})
+}
+
+type PayloadUpdater interface {
```
@dalehurwitz can you please unexport `PayloadUpdater` to `payloadUpdater` and `UpdatePayload` to `updatePayload`. Double check that it works with them unexported but pretty sure it does.

jonathaningram replied:
@dalehurwitz also `UpdatePayloadFunc` to `updatePayloadFunc`.

jonathaningram replied:
@dalehurwitz also `Payload` to `payload` and `PayloadItem` to `payloadItem` and also unexport all of:

```
UpdateEmailPayload
UpdatePayload
UpdateSMSPayload
```

Commit:
unexport private types and methods

jonathaningram commented on base/base.go:
```diff
@@ -0,0 +1,146 @@
+	Error    error
+}
+
+func (resp Response) JSON(v interface{}, at ...string) Response {
```
@dalehurwitz see platforms PR (will DM link). I'd actually like to unexport this base package by putting it inside an `internal` folder per the comment in the PR link above. If we decide later that we want to export it, it's easy to do.

jonathaningram commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+package main
```
@dalehurwitz remove this file?

Commit:
Move base package to internal folder

Commit:
Add docs for get all templates

Commit:
Improve error handling

jonathaningram committed:
Add checks to ensure valid API key

jonathaningram committed:
Don't export internal package struct field

jonathaningram committed:
Remove APIErrorFromError for now, and related docs

The function was exposing types from the internal base package so it was
not right. Will add in a later commit.
